src/mobu/services/github_ci/ci_manager.py

```diff
@@ -0,0 +1,336 @@
+        self._factory = GitHubAppClientFactory(
+            id=config.github_ci_app.id,
+            key=config.github_ci_app.private_key,
+            name="lsst-sqre/mobu CI app",
+            http_client=http_client,
+        )
+
+    async def start(self) -> None:
+        """Start the workers for the CI manager."""
+        self._logger.info("Starting CI manager...")
+        self.workers = [
+            Worker(
+                user=user,
+                queue=self._queue,
+                logger=self._logger,
+            )
+            for user in self._users
+        ]
+        self._logger.info("CI manager started")
+        self._jobs = [
+            await self._scheduler.spawn(worker.run())
+            for worker in self.workers
+        ]
+
+    async def aclose(self) -> None:
+        """Stop the workers and update GitHub CI checks for pending work.
+
+        We initially fail in-progress jobs too because we don't want any GitHub
+        check runs to be forever in-progress if mobu gets killed before they
+        finish. If the do end up finishing in time, then they'll be marked as
+        successful.
+
+        We'd rather have false-negative GitHub checks than forever-in-progress
+        checks, because:
+        * There is no way to re-run in-progress checks from the GitHub UI
+        * There is no way to know for sure from the GitHub UI that a check
+          will never be concluded.
+
+        A failed check can easily be re-run by any user, so this is an
+        acceptable tradeoff, assuming false-negatives happen infrequently.
+
+        Scenarios:
+
+        ### Job completes
+
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Job finishes before mobu is SIGKILLed
+        4. Mobu tells GitHub that job is successful
+        5. Mobu is SIGKILLed or exits cleanly
+
+        Result: GitHub check run displays success, which is corrcect.
+                An incorrect failure status will have been displayed for a
+                brief period of time.
+        User action needed: None
+
+        ### Job does not complete
+
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Mobu is SIGKILLed
+
+        Result: GitHub check run displays failure status, which is correct.
+        User action needed: Re-run check in GitHub UI
+
+        ### Job completes, but mobu is killed before it can tell GitHub
+
+        This situation should be pretty rare.
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Job finishes before mobu is SIGKILLed
+        4. Mobu is SIGKILLed before it can tell GitHub the job has succeeded
+
+        Result: GitHub check run displays failure status, which is incorrect.
+        User action needed: Re-run check in GitHub UI
+        """
+        self._logger.info("Stopping CI manager...")
+
```

Comment:
I'm guessing you're way ahead of me on this, but are we certain that all of the threads we're awaiting in the gather()s are going to complete in a finite time?  If not, do we want to put a timeout around the gather, or does that just leave us with stuff-hung-forever-on-GH?

Reply:
> are we certain that all of the threads we're awaiting in the gather()s are going to complete in a finite time?
We're absolutely not certain! At least we're not certain that the actual in-progress jobs will be finished in time (in fact, it's probably likely).  That in-progress work is awaited [here](https://github.com/lsst-sqre/mobu/blob/tickets%2FDM-44635%2Fmobu-in-ci/src/mobu/services/github_ci/ci_manager.py#L201-L203).

That's why we mark all of the in-progress (and queued) checks as failed as soon as we possibly can.  _That's_ what those gathers are doing, [here](https://github.com/lsst-sqre/mobu/blob/tickets%2FDM-44635%2Fmobu-in-ci/src/mobu/services/github_ci/ci_manager.py#L178-L185) and [here](https://github.com/lsst-sqre/mobu/blob/tickets%2FDM-44635%2Fmobu-in-ci/src/mobu/services/github_ci/ci_manager.py#L187-L194). I think we definitely want to wait as long as possible for those updates to complete because, right, if they don't, that leaves us with suff-hung-forever on GH. 

Hopefully there's enough time between the SIGTERM and the potential SIGKILL for those updates to happen. If there are other more robust ways to ensure no stuff-hung-forever, I'm totally open to anything.